3/debian-10/rootfs/opt/bitnami/scripts/etcd/prestop.sh

```
@@ -5,13 +5,18 @@
 set -o errexit
 set -o pipefail
 set -o nounset
+# set -o xtrace # Uncomment this line for debugging purposes
 
 # Load libraries
 . /opt/bitnami/scripts/libetcd.sh
 
 # Load etcd environment settings
 . /opt/bitnami/scripts/etcd-env.sh
 
+endpoints="$(etcdctl_get_endpoints true)"
```

PreStop hooks are executed **BEFORE** the TERM signal to stop the container is sent (see [Container Lifecycle docs](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/)). Therefore, why should we exclude the current member (which is running when this code is executed) from the list of endpoints to receive the "member remove" order?

---

I double checked the container lifecycle docs too and I currently have no good explanation for this behavior. The PreStop hook failed less often during my tests when the current member was excluded. Let me conduct some more tests tomorrow to see if I can see a pattern.

---

It's very weird.. Reading **etcd** docs (see https://etcd.io/docs/v3.3/op-guide/runtime-configuration/#remove-a-member) I'm pretty sure the member to be removed should also receive the "member remove" order...

There's only one explanation for your approach to fail less often that comes to my mind:

- The PreStop hook sends the "member remove" order to the current member that's being stopped (also the rest of members).
- The member is stopped since it receives the order BUT the hook hasn't finished yet since it's waiting for the order to be completed (receive the confirmation from every member).
- The hook finished abruptly (before completing the command) since the termination grace period expire, and the "member_removal.log" isn't created.

However, I can't confirm that's happening since it's not aligned with [Pod Termination docs](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination)...

We can try playing with the **terminationGracePeriodSeconds** to see if that helps.. I created [this PR](https://github.com/bitnami/charts/pull/7307) to add this feature to the chart.

---

> PreStop hooks are executed **BEFORE** the TERM signal to stop the container is sent (see [Container Lifecycle docs](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/)). Therefore, why should we exclude the current member (which is running when this code is executed) from the list of endpoints to receive the "member remove" order?

True, however when the script process fails early the POD gets into an Error > CrashLoopback state and it gets restarted, not terminated. What I've seen often happens is etcdctl adds the member then script fails and POD restarts, but now the member is already added and the subsequent error when add member runs again is cluster is in unhealthy state. IMO there should be some logic added to libetcd.sh script to either cleanup the member or continue from where it has left.